node/src/components/consensus/highway_core/state.rs

```diff
@@ -768,31 +862,76 @@ impl<C: Context> State<C> {
         if wunit.round_exp > self.params.max_round_exp() {
             return Err(UnitError::RoundLengthExpGreaterThanMaximum);
         }
-        if wunit.value.is_none() && !wunit.panorama.has_correct() {
+        if wunit.value.is_none()
+            && !wunit
+                .seq_num_panorama
+                .iter()
+                .any(ObservationSeqNum::is_correct)
+        {
             return Err(UnitError::MissingBlock);
         }
-        if wunit.panorama.len() != self.validator_count() {
-            return Err(UnitError::PanoramaLength(wunit.panorama.len()));
+        if wunit.seq_num_panorama.len() != self.validator_count() {
+            return Err(UnitError::PanoramaLength(wunit.seq_num_panorama.len()));
         }
-        if wunit.panorama.get(creator).is_faulty() {
-            return Err(UnitError::FaultyCreator);
+        match (wunit.seq_num_panorama.get(creator), wunit.previous) {
+            (ObservationSeqNum::Faulty, _) => return Err(UnitError::FaultyCreator),
+            (ObservationSeqNum::None, Some(_)) | (ObservationSeqNum::Correct(_), None) => {
+                return Err(UnitError::PreviousUnit);
+            }
+            (ObservationSeqNum::Correct(seq_num), Some(_)) => {
+                if seq_num.checked_add(1) != Some(wunit.seq_number) {
+                    return Err(UnitError::SequenceNumber);
+                }
+            }
+            (ObservationSeqNum::None, None) if wunit.seq_number != 0 => {
+                return Err(UnitError::SequenceNumber);
+            }
+            _ => {}
+        }
```

Comment:
Can you add a comment that we validate unit's `previous` against the panorama for consistency?